components/st-microelectronics/generate-ioc.stanza

```diff
@@ -0,0 +1,351 @@
+    if prefix?(mpn?(i) as String, "STM32") :
+      true
+  else :
+    false
+
+; defn add-comma-to-params (exist-params)
+;   if not exist-params :
+;     [""]
+;   else :
+;     [","]
+
+defn stm32-ip-name (ip : STM32IPs) -> String :
+  switch(ip) :
+    I2C1IP   : "I2C1" ; I2C1
+    I2C2IP   : "I2C2" ; I2C2
+    CAN1IP   : "CAN1" ; CAN1
+    CAN2IP   : "CAN2" ; CAN2
+    USART1IP : "USART1" ; USART1
+    USART2IP : "USART2" ; USART2
+    USART3IP : "USART3" ; USART3
+    UART4IP  : "UART4" ; UART4
+    UART5IP  : "UART5" ; UART5
+    USART6IP : "USART6" ; USART6
+    RCCIP    : "RCC" ; RCC - default
+    RTCIP    : "RTC" ; RTC - when adding 32kHz xtal?
+    SPI1IP   : "SPI1" ; SPI1
+    SPI2IP   : "SPI2" ; SPI2
+    SPI3IP   : "SPI3" ; SPI3
+    SYSIP    : "SYS" ; SYS - default
+    NVICIP   : "NVIC" ; NVIC - default
+    ADC1IP   : "ADC1" ; ADC1
+    ADC2IP   : "ADC2" ; ADC2
+    ADC3IP   : "ADC3" ; ADC3
+    DACIP    : "DAC" ; DAC
+
+
+defn stm32-spi-type (mode : SPIType) -> String :
+  switch(mode) :
+    SPIFullDuplexController : "Full_Duplex_Master"
+    SPIFullDuplexPeripheral : "Full_Duplex_Slave"
+    SPIHalfDuplexController : "Simplex_Bidirectional_Master"
+    SPIHalfDuplexPeripheral : "Simplex_Bidirectional_Slave"
+    SPIRxController :         "RX_Only_Simplex_Unidirect_Master"
+    SPIRxPeripheral :         "RX_Only_Simplex_Unidirect_Slave"
+    SPITxController :         "TX_Only_Simplex_Unidirect_Master"
+    SPITxPeripheral :         "TX_Only_Simplex_Unidirect_Slave"
+
+defn stm32-usart-mode (mode : USARTMode) -> String :
+  switch(mode) :
+    USARTSynchronous :    "Synchronous"
+    USARTAsynchronous :   "Asynchronous"
+    USARTSingleWire :     "Half_duplex(single_wire_mode)"
+    USARTMultiprocessor : "Multiprocessor_communication"
+    USARTIRDA :           "IrDA"
+    USARTSmartcard :      "SmartCard"
+    USARTSmartcardWClk :  "SmartCard_With_Clock"
+    USARTLIN :            "LIN"
+
+defn stm32-gpio-mode (mode : GPIOModes) -> String :
+  switch(mode) :
+    GPIOOutput :   "GPIO_Output"
+    GPIOInput :    "GPIO_Input"
+    GPIOAnalog :   "GPIO_Analog"
+    GPIOEventOut : "EVENTOUT"
+    GPIOExtI :     "GPXTI\n# can't currently write EXTI pins correctly."
+
+defn stm32-io-speed (speed : GPIOSpeed) -> String :
+  switch(speed) :
+    GPIOSpeedLow :      "GPIO_SPEED_FREQ_LOW"
+    GPIOSpeedMedium :   "GPIO_SPEED_FREQ_MEDIUM"
+    GPIOSpeedHigh :     "GPIO_SPEED_FREQ_HIGH"
+    GPIOSpeedVeryHigh : "GPIO_SPEED_FREQ_VERY_HIGH"
+
+defn stm32-pupd (mode : GPIOPuPd) -> String :
+  switch(mode) :
+    GPIOPullUp :   "GPIO_PULLUP"
+    GPIOPullDown : "GPIO_PULLDOWN"
+    GPIONoPull :   "GPIO_NOPULL"
+
+defn compose-ioc-contents (i : JITXObject) :
+  val ip-template = "Mcu.IP%_=%_\n"
+  val pin-template = "Mcu.Pin%_=%_\n"
+  var content : String = ""
+  var pin-content : String = ""
+  var pin-params : String = ""
+  var ip-content : String = ""
+  val ip-names = HashSet<STM32IPs>()
+  var exist-params = false
```

Comment:
I suggest naming this `exist-params?` to match Standard Stanza for boolean variable names.